### PR TITLE
[Backend] As a User, my uploaded keywords should be processed immediately

### PR DESCRIPTION
**Pull request by Goose97**
[[Backend] As a User, my uploaded keywords should be processed immediately](https://github.com/Goose97/google-scraper-ruby/issues/15)

## What happened 👀

Integrate keyword scraping with background job system backed by Sidekiq

## Insight 📝

- Jobs are enqueued immediately after keywords are created via `after_create_commit` callback
- The retry mechanism is currently implemented using ActiveJob `retry_on` method. Another alternative that I've considered is using Sidekiq native retry mechanism. I chose ActiveJob because:
  - After looked up others IC, it seems like we prefer using ActiveJob
  - Handling retry times exhaustion (in this case, we set keyword status to `:failed`) with ActiveJob is more convenient. Sidekiq has an option to do that, but ActiveJob is [yet to support it](https://github.com/sidekiq/sidekiq/pull/5994)
  
~~There is one caveat when choosing ActiveJob. My hunch is that if the Ruby node is restarted when a job is processing/retrying, the job will be marked as failed and since [we disable Sidekiq retry](https://github.com/Goose97/google-scraper-ruby/blob/6ea9f10725679d8fa43749ab837bae6754cbee43/app/jobs/scrape_keyword_job.rb#L6), the job will be lost. I'm looking for documentations to confirm this. Can I have your words on this @andyduong1920 ?~~

My bad, I thought that background worker threads are run inside Rails server process. They run in Sidekiq process, so restart the server process won't affect running jobs

### CODE REVIEW

**Comment by Goose97** [comment]
There is still one test left to be implemented: ScrapeKeyword jobs get enqueued after CSV file is uploaded. It's waiting for [this PR](https://github.com/Goose97/google-scraper-ruby/pull/31) to be merged

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (Goose97) [inline spec/jobs/scrape_keyword_job_spec.rb]: This test makes sure the background job worker handles errors correctly. The above codes is a workaround for `allow_any_instance_of` since that one is not recommended by the community.

However, this is just a hack and basically still do the same thing. 

What is your opinion on this?

File: app/models/keyword.rb
Comment (andyduong1920) [inline app/models/keyword.rb:19]:
```suggestion
  private
  
  def enqueue_scrape_job
```

File: app/jobs/scrape_keyword_job.rb
Comment (andyduong1920) [inline app/jobs/scrape_keyword_job.rb:13]: how about we set it as failed right in the `perform` action 🤔 

```
def perform(keyword_id:)
    Keyword.update(keyword_id, status: :processing)
    Google::ScrapeService.new(keyword_id: keyword_id).call!
  rescue GoogleScraperRuby::Errors::ScrapeError => error
    Keyword.update(keyword_id, status: :failed)
    raise error unless error.kind == :invalid_keyword
  end
```

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (andyduong1920) [reply spec/jobs/scrape_keyword_job_spec.rb]: This is normal @Goose97, I don't see any issue with that :)

File: app/models/keyword.rb
Comment (andyduong1920) [inline app/models/keyword.rb:15]: did you check this on your local, technically this branch need to be checkout from the `https://github.com/Goose97/google-scraper-ruby/pull/31` so you could test the whole flow there..

Comment (andyduong1920) [inline app/models/keyword.rb:15]: We are also missing the unit test for this part - after creating a keyword, it enqueues the job

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (Goose97) [reply spec/jobs/scrape_keyword_job_spec.rb]: cool. I just realize that, while the overall behavior is the same, the `allow_any_instance_of ` is more intrusive. It also stubs existing instances, whereas the above code only stubs newly created instances. TIL

File: app/models/keyword.rb
Comment (Goose97) [reply app/models/keyword.rb:15]: Yep, I left a note on this PR. Once the #31 is merged, this test will be implemented

Comment (Goose97) [reply app/models/keyword.rb:15]: yeah, I tested against #31 in my local setup. This PR depends on #31, hence the draft status. Once #31 is merged, I will have the whole flow tested again

File: app/jobs/scrape_keyword_job.rb
Comment (Goose97) [reply app/jobs/scrape_keyword_job.rb:13]: Oh, I see your point. So the job will be seen as `:failed` in between retry attempts, and marked as `:processing` before each attempt.

My current goal is: only marks a job as `:failed ` if it exceeds the maximum retries count (which is 3). So during the retry period, the status is still `:processing`. This is not an AC so I guess we can relax that.

But even if we go with your suggestion, the issue is still there, am I right? If the update line crashes, there no retry mechanism left to catch the job and it go straight to Sidekiq's dead queue. Anw, this is not a serious issue, I think we can left it like this

Comment (olivierobert) [reply app/jobs/scrape_keyword_job.rb:13]: It is indeed not a major concern as the UI could have a retry mechanism (out of scope though for the IC). Always aim for covering the happy paths first. Further improvements for edge cases can be handled in subsequent stories.

File: app/models/keyword.rb
Comment (andyduong1920) [reply app/models/keyword.rb:15]: Next time instead of waiting for the #31 is merged, you can checkout from there directly to have the full flow, what do you think about that idea?

Comment (Goose97) [reply app/models/keyword.rb:15]: wouldn't it make make code reviewing much harder, since we can not isolate changes from this PR 🤔 ?

Comment (andyduong1920) [reply app/models/keyword.rb:15]: No @Goose97 as long as the base branch point to the correct branch which is #31, this PR will display only its changes

image

Comment (Goose97) [reply app/models/keyword.rb:15]: Added test in 32de9e2b3f09716e7fb500a9a6bf522263a955b7

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (olivierobert) [inline spec/jobs/scrape_keyword_job_spec.rb]: MAybe missing parentheses for code consistency?

```suggestion
          assert_performed_jobs(1)
```

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (olivierobert) [inline spec/jobs/scrape_keyword_job_spec.rb]: This syntax can work, but the team usually uses the following to set up stubs/mocks:

```suggestion
          scrape_service = instance_double (Google::ScrapeService)
```

**Review by andyduong1920 — APPROVED** [review]

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (Goose97) [reply spec/jobs/scrape_keyword_job_spec.rb]: Fixed in 810d359d3574f9595bdc925c5eb031821d37138a

Not sure why Rubocop lets this passes

File: spec/jobs/scrape_keyword_job_spec.rb
Comment (Goose97) [reply spec/jobs/scrape_keyword_job_spec.rb]: Fixed in 57151f445a6060b23ed73c52de53e51af407d0b9

**Review by olivierobert — APPROVED** [review]

**Pull request merged**
